src/saml.py

```diff
@@ -2,10 +2,12 @@
 # See LICENSE file for licensing details.
 
 """Provide the SamlApp class to encapsulate the business logic."""
+import base64
+import hashlib
 import logging
 import urllib.request
 from functools import cached_property
-from typing import List, Set
+from typing import List, Optional
```

Comment:
I'm using `typing.List` because I haven't figured out why the built in types are failing here https://github.com/canonical/saml-integrator-operator/pull/17
Regarding `tyiping.Optional`, I get a `TypeError: unsupported operand type(s) for |: 'str' and 'NoneType` for `def signature(self) -> "etree.ElementTree" | None`